File: 2018/python/day22.py
#!/usr/bin/env python3
import time
from enum import *
from functools import *
from itertools import *
from operator import *

import networkx


class ToolType(Enum):
    TORCH = 0
    CLIMBING_GEAR = 1
    NEITHER = 2

class RegionType(Enum):
    ROCKY = 0
    WET = 1
    NARROW = 2

def solve(depth, dstX, dstY):
    @lru_cache(maxsize=None)
    def erosion_level(x, y):
        return (geologic_index(x, y) + depth) % 20183

    @lru_cache(maxsize=None)
    def geologic_index(x, y):
        if x == 0 and y == 0:
            return 0
        if x == dstX and y == dstY:
            return 0
        if x == 0:
            return y * 48271
        if y == 0:
            return x * 16807
        return erosion_level(x - 1, y) * erosion_level(x, y - 1)

    def region_type(x, y):
        return RegionType(erosion_level(x, y) % 3)

    print("pt1", sum(region_type(x, y).value for x, y in product(range(dstX+1), range(dstY+1))))

    # TIL A* algorithm
    # https://en.wikipedia.org/wiki/A*_search_algorithm

    # Construct a graph with edges between tiles for tools that
    # also works on neighboring tiles, and self-looping edges
    # for tool switching.

    t0 = time.time()
    COMPAT_GEAR = {
        RegionType.ROCKY: set([ToolType.CLIMBING_GEAR, ToolType.TORCH]),
        RegionType.WET: set([ToolType.CLIMBING_GEAR, ToolType.NEITHER]),
        RegionType.NARROW: set([ToolType.TORCH, ToolType.NEITHER])
    }

    def add_gear_switching(graph, typ, x, y):
        gear1, gear2 = COMPAT_GEAR[typ]
        graph.add_edge((x, y, gear1), (x, y, gear2), weight=7)

    def neighbors(x, y):
        if x != 0:
            yield (x - 1, y)
        yield (x + 1, y)
        if y != 0:
            yield (x, y - 1)
        yield (x, y + 1)

    def add_neighbors(graph, typ, x, y):
        gear1, gear2 = COMPAT_GEAR[typ]
        for nx, ny in neighbors(x, y):
            next_gear = COMPAT_GEAR[region_type(nx, ny)]
            if gear1 in next_gear:
                graph.add_edge((x, y, gear1), (nx, ny, gear1), weight=1)
            if gear2 in next_gear:
                graph.add_edge((x, y, gear2), (nx, ny, gear2), weight=1)

    graph = networkx.Graph()
    for x, y in product(range(dstX+30), range(dstY+1)):
        typ = region_type(x, y)
        add_gear_switching(graph, typ, x, y)
        add_neighbors(graph, typ, x, y)

    def manhattan(a, b):
        x1, y1, _ = a
        x2, y2, _ = b
        return abs(x2 - x1) + abs(y2 - y1)

    print("pt2", networkx.algorithms.astar_path_length(graph, (0, 0, ToolType.TORCH), (dstX, dstY, ToolType.TORCH), manhattan))


#solve(510, 10, 10)  #   114, 45
solve(4848, 15, 700) # 11359, 976

File: 2018/haskell/src/Day22.hs
{-# LANGUAGE LambdaCase        #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RecordWildCards   #-}
module Day22 (modeMazePt1, modeMazePt2) where

import           Algorithm.Search                 (aStar)
import           Control.Monad                    (guard)
import qualified Data.Attoparsec.ByteString.Char8 as P
import qualified Data.ByteString.Char8            as BS
import           Data.Maybe                       (fromJust)
import           Data.MemoTrie                    (memo2)

type RegionTypeSolver = Coord -> RegionType

data RegionType = Rocky | Wet | Narrow deriving (Enum, Show)
data ToolType = ClimbingGear | Torch | Neither deriving (Eq, Ord, Show)

type Depth = Int
type Coord = (Int, Int)

data State = State
 { stateCoord :: Coord
 , stateTool  :: ToolType } deriving (Eq, Ord, Show)

readInput :: String -> (Depth, Coord)
readInput = unwrap . P.parseOnly parser . BS.pack
  where unwrap (Right v) = v
        parser = do
          P.string "depth: "
          depth <- P.many1 P.digit
          P.endOfLine
          P.string "target: "
          dstX <- P.many1 P.digit
          P.char ','
          dstY <- P.many1 P.digit
          pure (read depth, (read dstX, read dstY))

genRegionTypeSolver :: Depth -> Coord -> RegionTypeSolver
genRegionTypeSolver depth (dstX, dstY) = regionType
  where regionType (x, y) = toEnum $ erosionLevel x y `mod` 3
        erosionLevel posX posY = (geologicIndex' posX posY + depth) `mod` 20183
        geologicIndex' = memo2 geologicIndex
        geologicIndex posX posY
          | posX == 0    && posY == 0    = 0
          | posX == dstX && posY == dstY = 0
          | posX == 0                    = posY * 48271
          | posY == 0                    = posX * 16807
          | otherwise                    = erosionLevel (posX - 1) posY *
                                           erosionLevel posX (posY - 1)

modeMazePt1 :: String -> Int
modeMazePt1 input = sum $ map regionRisk [(pX,pY) | pX <- [0..dstX], pY <- [0..dstY]]
  where (depth, target@(dstX, dstY)) = readInput input
        regionType = genRegionTypeSolver depth target
        regionRisk = fromEnum . regionType

compatibleTools :: RegionType -> [ToolType]
compatibleTools = \case
  Rocky  -> [ClimbingGear, Torch  ]
  Wet    -> [ClimbingGear, Neither]
  Narrow -> [Torch,        Neither]

possibleTools :: RegionTypeSolver -> Coord -> [ToolType]
possibleTools regionType = compatibleTools . regionType

possibleCoords :: Coord -> [Coord]
possibleCoords (x, y) = [(x', y') | (x', y') <- [(x - 1, y), (x + 1, y), (x, y - 1), (x, y + 1)], x' >= 0, y' >= 0]

neighbors :: (Coord -> [ToolType]) -> State -> [State]
neighbors possibleTools State{..} = do
  coord <- possibleCoords stateCoord
  tool <- possibleTools coord
  guard (tool `elem` validTools)
  pure (State coord tool)
  where validTools = possibleTools stateCoord

cost :: State -> State -> Int
cost State{stateTool=t1} State{stateTool=t2} = if t1 == t2 then 1 else 7 + 1

manhattan :: Coord -> State -> Int
manhattan c1 State{stateCoord=c2} = go c1 c2
  where go (x1,y1) (x2,y2) = abs (x1 - x2) + abs (y1 - y2)

modeMazePt2 :: String -> Int
modeMazePt2 input = fst . fromJust $ aStar candidates cost remainingCost isTargetState initialState
  where (depth, target@(dstX, dstY)) = readInput input
        regionType = genRegionTypeSolver depth target
        candidates = neighbors $ possibleTools regionType
        remainingCost = manhattan target
        initialState = State (0, 0) Torch
        isTargetState = (== State target Torch)
